File: mobileapp/lib/features/home/presentation/utils/festival_theme_resolver.dart
import 'package:flutter/material.dart';
import '../../data/models/festival_campaign_model.dart';

class ResolvedFestivalTheme {
  final String key;
  final String emoji;
  final String fontPreset;
  final LinearGradient backgroundGradient;
  final Color backgroundColor;
  final Color cardBackground;
  final Color cardBorder;
  final Color accentColor;
  final Color buttonColor;
  final Color textColor;

  /// Readable colour for text/icons sitting directly on [cardBackground]
  /// (i.e. not over a dark image overlay). White on a dark card, dark ink on a
  /// light one — so an admin-picked pastel card background never gets
  /// unreadable white text. Mirrors the web resolver's `cardText`.
  final Color cardText;

  const ResolvedFestivalTheme({
    required this.key,
    required this.emoji,
    required this.fontPreset,
    required this.backgroundGradient,
    required this.backgroundColor,
    required this.cardBackground,
    required this.cardBorder,
    required this.accentColor,
    required this.buttonColor,
    required this.textColor,
    required this.cardText,
  });
}

/// Rough perceived-luminance test so text stays readable whatever colour the
/// admin picked. Mirrors `isDarkColor` in the web `festivalThemeResolver.ts`.
bool isDarkColor(Color c) {
  final luminance = 0.2126 * c.red + 0.7152 * c.green + 0.0722 * c.blue;
  return luminance < 140;
}

class FestivalThemeResolver {
  static const Map<String, Map<String, dynamic>> _presets = {
    'krishna': {
      'emoji': '🦚',
      'fontPreset': 'greatVibes',
      'gStart': Color(0xFF8DC9F7), // Soft sky blue at location level (Image 2)
      'gEnd': Color(0xFFD6F0FE),   // Soft light pastel body blue
      'cardBg': Color(0xFF8DC9F7), // Matches location section background color!
      'cardBorder': Color(0xFFBAE6FD),
      'accent': Color(0xFFD97706),
      'btn': Color(0xFF0284C7),
      'text': Color(0xFF0F4C75),
    },
    'ganesh_chaturthi': {
      'emoji': '🌺',
      'fontPreset': 'rozhaOne',
      'gStart': Color(0xFFFCDAA8), // Soft warm peach at location level (Image 3)
      'gEnd': Color(0xFFFEF3E2),   // Soft light warm cream body
      'cardBg': Color(0xFFFCDAA8), // Matches location section background color!
      'cardBorder': Color(0xFFFDE68A),
      'accent': Color(0xFFD97706),
      'btn': Color(0xFFEA580C),
      'text': Color(0xFF68380D),
    },
    'diwali': {
      'emoji': '🪔',
      'fontPreset': 'rozhaOne',
      'gStart': Color(0xFFFCD39D),
      'gEnd': Color(0xFFFEF4E6),
      'cardBg': Color(0xFFFCD39D), // Matches location section background color!
      'cardBorder': Color(0xFFFDE68A),
      'accent': Color(0xFFD97706),
      'btn': Color(0xFFC2410C),
      'text': Color(0xFF663000),
    },
    'onam': {
      'emoji': '🌸',
      'fontPreset': 'cinzelDecorative',
      'gStart': Color(0xFFBBEB9B),
      'gEnd': Color(0xFFEBFADF),
      'cardBg': Color(0xFFBBEB9B), // Matches location section background color!
      'cardBorder': Color(0xFFC0F289),
      'accent': Color(0xFFD97706),
      'btn': Color(0xFF16A34A),
      'text': Color(0xFF1B4D20),
    },
    'raksha_bandhan': {
      'emoji': '🧿',
      'fontPreset': 'satisfy',
      'gStart': Color(0xFFF8A6D2),
      'gEnd': Color(0xFFFCE6F2),
      'cardBg': Color(0xFFF8A6D2), // Matches location section background color!
      'cardBorder': Color(0xFFFBCFE8),
      'accent': Color(0xFFEC4899),
      'btn': Color(0xFF9333EA),
      'text': Color(0xFF701A75),
    },
    'holi': {
      'emoji': '🎨',
      'fontPreset': 'pacifico',
      'gStart': Color(0xFFA5B4FC),
      'gEnd': Color(0xFFE0E7FF),
      'cardBg': Color(0xFFA5B4FC), // Matches location section background color!
      'cardBorder': Color(0xFFC7D2FE),
      'accent': Color(0xFFE11D48),
      'btn': Color(0xFF4F46E5),
      'text': Color(0xFF1E1B4B),
    },
    'navratri': {
      'emoji': '🪷',
      'fontPreset': 'cinzelDecorative',
      'gStart': Color(0xFFD8B4FE),
      'gEnd': Color(0xFFF3E8FF),
      'cardBg': Color(0xFFD8B4FE), // Matches location section background color!
      'cardBorder': Color(0xFFE9D5FF),
      'accent': Color(0xFF9333EA),
      'btn': Color(0xFF7E22CE),
      'text': Color(0xFF4C1D95),
    },
  };

  static ResolvedFestivalTheme resolve(FestivalCampaignModel campaign) {
    final key = campaign.themeKey.toLowerCase();
    final preset = _presets[key] ?? _presets['krishna']!;

    Color gStart = preset['gStart'];
    Color gEnd = preset['gEnd'];
    Color solid = preset['gStart'];

    if (campaign.backgroundType == 'solid') {
      solid = campaign.backgroundColor;
      gStart = solid;
      gEnd = solid;
    } else if (campaign.backgroundType == 'gradient') {
      gStart = campaign.gradientStart;
      gEnd = campaign.gradientEnd;
      solid = gStart;
    }

    final styling = campaign.cardStyling;

    Alignment begin = Alignment.topCenter;
    Alignment end = Alignment.bottomCenter;
    final dir = campaign.gradientDirection.toLowerCase();
    if (dir.contains('diagonal') || dir.contains('top left') || dir.contains('135') || dir.contains('\\')) {
      begin = Alignment.topLeft;
      end = Alignment.bottomRight;
    } else if (dir.contains('right') || dir.contains('horizontal')) {
      begin = Alignment.centerLeft;
      end = Alignment.centerRight;
    }

    final cardBg = gStart;

    // Content on the card must read against whatever colour the admin chose:
    // white on a dark card, else the admin text colour if it is itself dark,
    // else a near-black ink.
    final cardText = isDarkColor(cardBg)
        ? Colors.white
        : (isDarkColor(styling.textColor) ? styling.textColor : const Color(0xFF1C1C1E));

    return ResolvedFestivalTheme(
      key: key,
      emoji: preset['emoji'],
      fontPreset: preset['fontPreset'] ?? 'greatVibes',
      backgroundGradient: LinearGradient(
        colors: [gStart, gEnd],
        begin: begin,
        end: end,
      ),
      backgroundColor: solid,
      cardBackground: cardBg,
      cardBorder: styling.cardBorder,
      accentColor: styling.accentColor,
      buttonColor: styling.buttonColor,
      textColor: styling.textColor,
      cardText: cardText,
    );
  }
}

File: frontend/src/utils/festivalThemeResolver.ts
import { FestivalCampaign } from '../context/CMSContext';

export interface ResolvedFestivalTheme {
  key: string;
  emoji: string;
  fontPreset: string;
  /** Ready-to-use CSS font-family stack for the festival heading. */
  fontFamily: string;
  gStart: string;
  gEnd: string;
  bgGradient: string;
  cardBg: string;
  cardBorder: string;
  /** Readable text colour for content sitting directly on `cardBg`. */
  cardText: string;
  accent: string;
  btn: string;
  text: string;
}

/** Rough perceived-luminance test so text on a card stays readable
 *  whatever colour the admin picked. */
export function isDarkColor(hex?: string): boolean {
  if (!hex || typeof hex !== 'string') return false;
  let h = hex.replace('#', '').trim();
  if (h.length === 3) h = h.split('').map((c) => c + c).join('');
  if (h.length !== 6) return false;
  const r = parseInt(h.slice(0, 2), 16);
  const g = parseInt(h.slice(2, 4), 16);
  const b = parseInt(h.slice(4, 6), 16);
  return (0.2126 * r + 0.7152 * g + 0.0722 * b) < 140;
}

const FONT_STACKS: Record<string, string> = {
  greatVibes: "'Great Vibes', 'Brush Script MT', cursive",
  rozhaOne: "'Rozha One', Georgia, 'Times New Roman', serif",
  cinzelDecorative: "'Cinzel Decorative', Georgia, serif",
  satisfy: "'Satisfy', 'Brush Script MT', cursive",
  pacifico: "'Pacifico', 'Comic Sans MS', cursive",
};

const PRESETS: Record<string, Omit<ResolvedFestivalTheme, 'cardText' | 'fontFamily'>> = {
  krishna: {
    key: 'krishna',
    emoji: '🦚',
    fontPreset: 'greatVibes',
    gStart: '#8DC9F7',
    gEnd: '#D6F0FE',
    bgGradient: 'linear-gradient(to bottom, #8DC9F7, #D6F0FE)',
    cardBg: '#8DC9F7',
    cardBorder: '#BAE6FD',
    accent: '#D97706',
    btn: '#0284C7',
    text: '#0F4C75',
  },
  ganesh_chaturthi: {
    key: 'ganesh_chaturthi',
    emoji: '🌺',
    fontPreset: 'rozhaOne',
    gStart: '#FCDAA8',
    gEnd: '#FEF3E2',
    bgGradient: 'linear-gradient(to bottom, #FCDAA8, #FEF3E2)',
    cardBg: '#FCDAA8',
    cardBorder: '#FDE68A',
    accent: '#D97706',
    btn: '#EA580C',
    text: '#68380D',
  },
  diwali: {
    key: 'diwali',
    emoji: '🪔',
    fontPreset: 'rozhaOne',
    gStart: '#FCD39D',
    gEnd: '#FEF4E6',
    bgGradient: 'linear-gradient(to bottom, #FCD39D, #FEF4E6)',
    cardBg: '#FCD39D',
    cardBorder: '#FDE68A',
    accent: '#D97706',
    btn: '#C2410C',
    text: '#663000',
  },
  onam: {
    key: 'onam',
    emoji: '🌸',
    fontPreset: 'cinzelDecorative',
    gStart: '#BBEB9B',
    gEnd: '#EBFADF',
    bgGradient: 'linear-gradient(to bottom, #BBEB9B, #EBFADF)',
    cardBg: '#BBEB9B',
    cardBorder: '#C0F289',
    accent: '#D97706',
    btn: '#16A34A',
    text: '#1B4D20',
  },
  raksha_bandhan: {
    key: 'raksha_bandhan',
    emoji: '🧿',
    fontPreset: 'satisfy',
    gStart: '#F8A6D2',
    gEnd: '#FCE6F2',
    bgGradient: 'linear-gradient(to bottom, #F8A6D2, #FCE6F2)',
    cardBg: '#F8A6D2',
    cardBorder: '#FBCFE8',
    accent: '#EC4899',
    btn: '#9333EA',
    text: '#701A75',
  },
  holi: {
    key: 'holi',
    emoji: '🎨',
    fontPreset: 'pacifico',
    gStart: '#A5B4FC',
    gEnd: '#E0E7FF',
    bgGradient: 'linear-gradient(to bottom, #A5B4FC, #E0E7FF)',
    cardBg: '#A5B4FC',
    cardBorder: '#C7D2FE',
    accent: '#E11D48',
    btn: '#4F46E5',
    text: '#1E1B4B',
  },
  navratri: {
    key: 'navratri',
    emoji: '🪷',
    fontPreset: 'cinzelDecorative',
    gStart: '#D8B4FE',
    gEnd: '#F3E8FF',
    bgGradient: 'linear-gradient(to bottom, #D8B4FE, #F3E8FF)',
    cardBg: '#D8B4FE',
    cardBorder: '#E9D5FF',
    accent: '#9333EA',
    btn: '#7E22CE',
    text: '#4C1D95',
  },
};

export function resolveFestivalTheme(campaign?: FestivalCampaign | null): ResolvedFestivalTheme {
  if (!campaign) {
    return {
      ...PRESETS.krishna,
      cardText: isDarkColor(PRESETS.krishna.cardBg) ? '#FFFFFF' : '#1C1C1E',
      fontFamily: FONT_STACKS[PRESETS.krishna.fontPreset] || 'Georgia, serif',
    };
  }

  const key = (campaign.themeKey || 'krishna').toLowerCase();
  const preset = PRESETS[key] || PRESETS.krishna;

  // Resolve Admin Background Colors & Gradient
  const rawBgColor =
    (typeof campaign.theme === 'string'
      ? campaign.theme
      : campaign.theme?.backgroundColor || campaign.theme?.themeColor || campaign.theme?.color) ||
    campaign.backgroundColor;

  let gStart = campaign.gradientStart || rawBgColor || preset.gStart;
  let gEnd = campaign.gradientEnd || campaign.gradientStart || rawBgColor || preset.gEnd;

  if (campaign.backgroundType === 'solid' && (rawBgColor || campaign.gradientStart)) {
    const solidColor = rawBgColor || campaign.gradientStart || preset.gStart;
    gStart = solidColor;
    gEnd = solidColor;
  }

  let direction = 'to bottom';
  if (campaign.gradientDirection) {
    const dir = campaign.gradientDirection.toLowerCase();
    if (dir.includes('diagonal') || dir.includes('top left') || dir.includes('135') || dir.includes('\\')) {
      direction = '135deg';
    } else if (dir.includes('right') || dir.includes('horizontal')) {
      direction = 'to right';
    }
  }

  const bgGradient = gStart === gEnd ? gStart : `linear-gradient(${direction}, ${gStart}, ${gEnd})`;

  // Resolve Card Styling Tokens (prioritizing Admin input -> top-level -> preset)
  const styling = campaign.cardStyling || {};
  const cardBg =
    styling.cardBackground ||
    (campaign as any).cardBackground ||
    (campaign.backgroundType === 'solid' ? gStart : null) ||
    gStart;
  const cardBorder = styling.cardBorder || (campaign as any).cardBorder || preset.cardBorder;
  const accent = styling.accentColor || (campaign as any).accentColor || preset.accent;
  const btn = styling.buttonColor || (campaign as any).buttonColor || preset.btn;
  const text = styling.textColor || (campaign as any).textColor || preset.text;

  // Content on the card must read against whatever colour the admin chose.
  const cardText = isDarkColor(cardBg) ? '#FFFFFF' : (isDarkColor(text) ? text : '#1C1C1E');

  const fontFamily =
    (campaign as any).fontFamily ||
    (styling as any).fontFamily ||
    FONT_STACKS[preset.fontPreset] ||
    'Georgia, serif';

  return {
    key,
    emoji: preset.emoji,
    fontPreset: preset.fontPreset,
    fontFamily,
    gStart,
    gEnd,
    bgGradient,
    cardBg,
    cardBorder,
    cardText,
    accent,
    btn,
    text,
  };
}
